src/protocols/router/hpr_router_connection_manager.erl

```diff
@@ -47,85 +46,66 @@ start_link() ->
 %% new connection if one doesn't exist.
 -spec get_connection(Lns :: lns()) -> {ok, grpc_client:connection()} | {error, any()}.
 get_connection(Lns) ->
-    DecodedLns = decode_lns(Lns),
-    gen_server:call(?MODULE, {get_connection, Lns, DecodedLns}).
+    case ets:lookup(?CONNECTION_TAB, Lns) of
+        [ConnectionRecord] ->
+            {ok, ConnectionRecord#connection.connection};
+        [] ->
+            gen_server:call(?MODULE, {connect, Lns})
+    end.
 
 %% ------------------------------------------------------------------
 %% gen_server Function Definitions
 %% ------------------------------------------------------------------
 
 -spec init([]) -> {ok, #state{}}.
 init([]) ->
-    Tab = init_ets(),
-    {
-        ok,
-        #state{
-            connection_table = Tab
-        }
-    }.
-
--spec handle_call({get_connection, lns(), endpoint()}, from(), #state{}) ->
-    {reply, {ok, grpc_client:connection()} | {error, any()}, #state{}}.
-handle_call({get_connection, Lns, Endpoint}, _From, State) ->
-    {reply, do_get_connection(Lns, Endpoint, State#state.connection_table), State}.
-
--spec handle_cast(Msg, #state{}) -> {stop, {unimplemented_cast, Msg}, #state{}}.
+    _ = ets:new(?CONNECTION_TAB, [public, named_table, set, {keypos, #connection.lns}]),
+    {ok, #state{}}.
+
+-spec handle_call(Msg :: any(), from(), #state{}) ->
+    {reply, {ok, grpc_client:connection()} | {error, any()}, #state{}}
+    | {stop, {unimplemented_call, any()}, #state{}}.
+handle_call({connect, Lns}, _From, State) ->
+    {Transport, Host, Port} = decode_lns(Lns),
+    case grpc_client:connect(Transport, Host, Port, []) of
+        {error, _} = Error ->
+            {reply, Error, State};
+        {ok, Conn} = OK ->
+            true = ets:insert(?CONNECTION_TAB, #connection{
+                lns = binary:copy(Lns),
+                connection = Conn
+            }),
+            #{http_connection := Pid} = Conn,
+            _ = erlang:monitor(process, Pid, [{tag, {'DOWN', Lns}}]),
+            {reply, OK, State}
+    end;
+handle_call(Msg, _From, State) ->
+    {stop, {unimplemented_call, Msg}, State}.
+
+-spec handle_cast(Msg :: any(), #state{}) ->
+    {noreply, #state{}} | {stop, {unimplemented_cast, any()}, #state{}}.
 handle_cast(Msg, State) ->
     {stop, {unimplemented_cast, Msg}, State}.
 
 -spec handle_info({{'DOWN', lns()}, reference(), process, pid(), any()}, #state{}) ->
     {noreply, #state{}}.
 handle_info({{'DOWN', Lns}, _Mon, process, _Pid, _ExitReason}, State) ->
-    ets:delete(State#state.connection_table, Lns),
+    lager:info("connection ~p to ~s went down ~p", [_Pid, Lns, _ExitReason]),
+    true = ets:delete(?CONNECTION_TAB, Lns),
+    {noreply, State};
+handle_info({'EXIT', _Pid, _Reason}, State) ->
+    lager:info("connection ~p exited ~p", [_Pid, _Reason]),
```

Comment:
I think it would be clearer if we add `process_flag(trap_exit, true)` to `init`, although I do see that `http2_client:new_connection` does this. I think that's a confusing side effect of creating a new `http2` connection, which makes it a little mysterious how we can get `'EXIT'` messages without setting the process flag

Reply:
Yes? But that we're getting one at all is kind of confusing if you don't know that `http2_client:new_connection` is setting the trap_exit process flag. I still think we should set it in `init`.